framework/base/Action.php

```diff
@@ -115,8 +132,97 @@ protected function beforeRun()
     /**
      * This method is called right after `run()` is executed.
      * You may override this method to do post-processing work for the action run.
+     * @return mixed the processed action result got from `$this->result`.
      */
     protected function afterRun()
     {
     }
+
+    /**
+     * Resolves action arguments, derived class my override this method to provide custom argument resolver
+     *
+     * @return array
+     */
+    public function resolveActionArguments(array $params)
+    {
+        $args = $this->controller->bindActionParams($this, $params);
+        if (Yii::$app->requestedParams === null) {
+            Yii::$app->requestedParams = $args;
+        }
+        return $args;
+    }
+
+    /**
+     * Executes action handler using resolved action arguments
+     *
+     * @param array $args the action handler arguments
+     * @return mixed the result of action handler invocation
+     */
+    protected function executeAction($args)
+    {
+        if (!method_exists($this, 'run')) {
+            throw new InvalidConfigException(get_class($this) . ' must define a "run()" method.');
+        }
+        
+        if ($this->beforeRun()) {
+            Yii::debug('Running action: ' . get_class($this) . '::run(), invoked by '  . get_class($this->controller), __METHOD__);
+            $this->result = call_user_func_array([$this, 'run'], $args);
+            $this->afterRun();
+        }        
+    }
+    
+    /**
+     * Get  argument passed to the specified parameter
+     *
+     * @param string $paramName name of declared parameter in inlineAction actionMethod/ Action::run().
+     * @return mixed value bound to the parameter `$paramName`
+     * @throws BadRequestHttpException if the parameter `$paramName` is not defined or argument not yet bound to it.
+     */
+    public function getRequestedParam($paramName)
```

Comment:
Can't we read Yii::$app->requestedParams instead of doing all the reflection 2nd time?

Reply:
Absolutely yes but, since values are mainly string and especially descriptive placeholders of  object injected, I didn't find a way of differentiating this string form an eventual string requested as it is (not as placeholders) that may have similar warding!

Any suggestion is appreciated.  

thanks

Reply:
I mean yes we may escape use of reflection but we need a kind of manipulation to translate [injectedparams ](https://github.com/yiisoft/yii2/blob/03befccd48a180ea85bf70f9cdf769c360e13adf/framework/web/Controller.php#L194) from placeholder string to argument instances

Reply:
Likely we can record type info into private requestedParamsMeta property when reflecting/filling requestedParams. That would make it one-time reflection.

Reply:
Sure, which will require get and set methods to access `requestedParamsMeta ` outside `\Yii::$app`. 
can I proceed this way?

Reply:
That's a bit hacky but is better performance-wise for such a common operation. @bizley what do you think?

Reply:
I think to maintain privacy of these injected prams,   the `requestedParamsMeta`  has to be read only through getter, which is only possible if it is defined as controller property where the `bindInjectedParams()` is defined as well.
defining it elsewhere with possibility of setting any paramValue out side even with setter will break its security.

My suggestion is in the last commit.

Thanks

Reply:
I agree with the private properties and getter/setter.

Reply:
setter I think should not, because a paramMeta info may be changed out side `bindInjectedParams()`!? isn't it a potential privacy break@bizley?   This info will have to be validated which requires use of reflection!!!

Reply:
Yes, probably no setter.

Reply:
Then `requestedParamsMeta` has to be read only and be declared where `bindInjectedParams()` is defined.